mayastor/src/host/device.rs

```diff
@@ -0,0 +1,327 @@
+//!  - the device is of an acceptable type as determined by well known device
+//!    numbers (eg. SCSI disks)
+//!  - the device represents either a disk with no partitions or a disk
+//!    partition of an acceptable type (Linux filesystem partitions only at
+//!    present)
+//!  - the device currently contains no filesystem or volume id (although this
+//!    logically implies that the device is not currently mounted, for the sake
+//!    of consistency, the mount table is also checked to ENSURE that the device
+//!    is not mounted)
+
+use std::{
+    collections::HashMap,
+    ffi::{OsStr, OsString},
+    io::Error,
+};
+
+use proc_mounts::{MountInfo, MountIter};
+use rpc::mayastor::{
+    block_device::{Filesystem, Partition},
+    BlockDevice,
+};
+use udev::{Device, Enumerator};
+
+// Struct representing a property value in a udev::Device struct (and possibly
+// elsewhere). It is used to provide conversions via various "From" trait
+// implementations below.
+struct Property<'a>(Option<&'a OsStr>);
+
+impl From<Property<'_>> for String {
+    fn from(property: Property) -> Self {
+        String::from(property.0.map(|s| s.to_str()).flatten().unwrap_or(""))
+    }
+}
+
+impl From<Property<'_>> for Option<String> {
+    fn from(property: Property) -> Self {
+        property.0.map(|s| s.to_str()).flatten().map(String::from)
+    }
+}
+
+impl From<Property<'_>> for Option<u32> {
+    fn from(property: Property) -> Self {
+        Option::<String>::from(property)
+            .map(|s| s.parse().ok())
+            .flatten()
+    }
+}
+
+impl From<Property<'_>> for u32 {
+    fn from(property: Property) -> Self {
+        Option::<Self>::from(property).unwrap_or(0)
+    }
+}
+
+impl From<Property<'_>> for Option<u64> {
+    fn from(property: Property) -> Self {
+        Option::<String>::from(property)
+            .map(|s| s.parse().ok())
+            .flatten()
+    }
+}
+
+impl From<Property<'_>> for u64 {
+    fn from(property: Property) -> Self {
+        Option::<Self>::from(property).unwrap_or(0)
+    }
+}
+
+// Determine the type of devices which may be potentially presented
+// as "available" for use.
+fn usable_device(devmajor: &u32) -> bool {
+    const DEVICE_TYPES: [u32; 4] = [
+        7,   // Loopback devices
+        8,   // SCSI disk devices
+        43,  // Network block devices
+        259, // Block Extended Major
+    ];
+
+    if DEVICE_TYPES.iter().any(|m| m == devmajor) {
+        return true;
+    }
+
+    // TODO: add extra logic here as needed for devices with dynamically
+    // allocated major numbers
+
+    false
+}
+
+// Determine the type of partitions which may be potentially presented
+// as "available" for use
+fn usable_partition(partition: &Option<Partition>) -> bool {
+    const GPT_PARTITION_TYPES: [&str; 1] = [
+        "0fc63daf-8483-4772-8e79-3d69d8477de4", // Linux
+    ];
+
+    const MBR_PARTITION_TYPES: [&str; 1] = [
+        "0x83", // Linux
+    ];
+
+    if let Some(part) = partition {
```

Comment:
you can write this as const ["UUID1", "UUID2"].iter().any(...) this avoids the if statements